lua/tint.lua

```diff
@@ -1,121 +1,147 @@
-  local r = math.floor(num / 0x10000)
-  local g = (math.floor(num / 0x100) % 0x100)
-  local b = (num % 0x100)
-  if saturation ~= 1 then
-    local rec601_luma = 0.299 * r + 0.587 * g + 0.114 * b
-    r = math.floor(r * saturation + rec601_luma * (1 - saturation))
-    g = math.floor(g * saturation + rec601_luma * (1 - saturation))
-    b = math.floor(b * saturation + rec601_luma * (1 - saturation))
-  end
+-- Default module configuration values
+__.default_config = {
+    tint = -40,
+    saturation = 0.7,
+    transforms = nil,
+    tint_background_colors = false,
+    highlight_ignore_patterns = {},
+    window_ignore_function = nil,
+}
 
-  return string.format("#%06x", clamp(r + amt) * 0x10000 + clamp(g + amt) * 0x100 + clamp(b + amt))
-end
+-- Pre-defined transforms that can be used by the user
+__.transforms = {
+  [tint.transforms.SATURATE_TINT] = function()
+    return {
+      colors.saturate(__.user_config.saturation),
+      colors.tint(__.user_config.tint),
+    }
+  end,
+}
 
-local function get_def(v)
-  -- Some keys in `v` are not valid - pass only valid ones here
+--- Ensure the passed table has only valid keys to hand to `nvim_set_hl`
+---
+---@param hl_def table Value returned by `nvim_get_hl_by_name` with `rgb` colors exported
+---@return table The passed highlight definition with valid `nvim_set_hl` keys only
+local function ensure_valid_hl_keys(hl_def)
   return {
-    fg = v.fg or v.foreground,
-    bg = v.bg or v.background,
-    sp = v.sp or v.special,
-    blend = v.blend,
-    bold = v.bold,
-    standout = v.standout,
-    underline = v.underline,
-    undercurl = v.undercurl,
-    underdouble = v.underdouble,
-    underdotted = v.underdotted,
-    underdashed = v.underdashed,
-    strikethrough = v.strikethrough,
-    italic = v.italic,
-    reverse = v.reverse,
-    nocombine = v.nocombine,
-    link = v.link,
-    default = v.default,
-    ctermfg = v.ctermfg,
-    ctermbg = v.ctermbg,
-    cterm = v.cterm,
+    fg = hl_def.fg or hl_def.foreground,
+    bg = hl_def.bg or hl_def.background,
+    sp = hl_def.sp or hl_def.special,
+    blend = hl_def.blend,
+    bold = hl_def.bold,
+    standout = hl_def.standout,
+    underline = hl_def.underline,
+    undercurl = hl_def.undercurl,
+    underdouble = hl_def.underdouble,
+    underdotted = hl_def.underdotted,
+    underdashed = hl_def.underdashed,
+    strikethrough = hl_def.strikethrough,
+    italic = hl_def.italic,
+    reverse = hl_def.reverse,
+    nocombine = hl_def.nocombine,
+    link = hl_def.link,
+    default = hl_def.default,
+    ctermfg = hl_def.ctermfg,
+    ctermbg = hl_def.ctermbg,
+    cterm = hl_def.cterm,
   }
 end
 
-local function ignore_tint(winid)
-  return tint.config.ignorefunc and tint.config.ignorefunc(winid) or false
+--- Get the set of transforms to apply to highlight groups from the colorscheme in question
+---
+---@return table A table of functions to transform the input RGB color values by
+local function get_transforms()
```

Comment:
This whole function might be something easier done once when the configuration is setup rather than each time `setup_all` is called (which only really happens once given the colorscheme is not changed).

Not the end of the world, but could be made a bit simpler.